crates/re_viewport/src/viewport_blueprint.rs

```diff
@@ -0,0 +1,294 @@
+}
+
+impl<'a> ViewportBlueprint<'a> {
+    pub fn from_db(blueprint_db: &'a re_data_store::StoreDb) -> Self {
+        let space_views: HashMap<SpaceViewId, SpaceViewBlueprint> = if let Some(space_views) =
+            blueprint_db
+                .entity_db
+                .tree
+                .children
+                .get(&re_data_store::EntityPathPart::Name(
+                    SpaceViewComponent::SPACEVIEW_PREFIX.into(),
+                )) {
+            space_views
+                .children
+                .values()
+                .filter_map(|view_tree| load_space_view(&view_tree.path, blueprint_db))
+                .map(|sv| (sv.id, sv))
+                .collect()
+        } else {
+            Default::default()
+        };
+
+        let viewport = load_viewport(blueprint_db, space_views);
+
+        Self {
+            blueprint_db,
+            viewport: viewport.clone(),
+            snapshot: viewport,
+        }
+    }
+
+    /// If `false`, the item is referring to data that is not present in this blueprint.
+    pub fn is_item_valid(&self, item: &Item) -> bool {
+        match item {
+            Item::ComponentPath(_) => true,
+            Item::InstancePath(space_view_id, _) => space_view_id
+                .map(|space_view_id| self.viewport.space_view(&space_view_id).is_some())
+                .unwrap_or(true),
+            Item::SpaceView(space_view_id) => self.viewport.space_view(space_view_id).is_some(),
+            Item::DataBlueprintGroup(space_view_id, data_blueprint_group_handle) => {
+                if let Some(space_view) = self.viewport.space_view(space_view_id) {
+                    space_view
+                        .data_blueprint
+                        .group(*data_blueprint_group_handle)
+                        .is_some()
+                } else {
+                    false
+                }
+            }
+        }
+    }
+
+    pub fn sync_changes(&self, command_sender: &CommandSender) {
+        let mut deltas = vec![];
+
+        sync_viewport(&mut deltas, &self.viewport, &self.snapshot);
+
+        // Add any new or modified space views
+        for id in self.viewport.space_view_ids() {
+            if let Some(space_view) = self.viewport.space_view(id) {
+                sync_space_view(&mut deltas, space_view, self.snapshot.space_view(id));
+            }
+        }
+
+        // Remove any deleted space views
+        for space_view_id in self.snapshot.space_view_ids() {
+            if self.viewport.space_view(space_view_id).is_none() {
+                clear_space_view(&mut deltas, space_view_id);
+            }
+        }
+
+        command_sender.send_system(SystemCommand::UpdateBlueprint(
+            self.blueprint_db.store_id().clone(),
+            deltas,
+        ));
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+// TODO(jleibs): Move this helper to a better location
+fn push_one_component<C: SerializableComponent>(
```

Comment:
I find this name very confusing, my intuition was that this would add a component to an existing row.